[Bug] Partition type filter error when insert overwrite statement use date or time type

### Search before asking

- [X] I searched in the [issues](https://github.com/apache/paimon/issues) and found nothing similar.


### Paimon version

Paimon 0.8 snapshot

### Compute Engine

Spark & Flink

### Minimal reproduce step

```
CREATE TABLE if not exists paimon.paimon.test_partition (
  id STRING COMMENT '租户ID',
  dt date)
USING paimon
PARTITIONED BY (dt)
TBLPROPERTIES (
  'primary-key' = 'id,dt',
 'bucket' = '3'
  );
```
```
insert overwrite  paimon.paimon.test_partition partition (dt='2024-04-18') values(1)
```

```
java.lang.RuntimeException: java.time.DateTimeException: For input string: '19831'.
	at org.apache.paimon.spark.commands.PaimonSparkWriter.commit(PaimonSparkWriter.scala:111)
	at org.apache.paimon.spark.commands.WriteIntoPaimonTable.run(WriteIntoPaimonTable.scala:64)
	at org.apache.paimon.spark.SparkWrite.$anonfun$toInsertableRelation$1(SparkWrite.scala:35)
	at org.apache.spark.sql.execution.datasources.v2.SupportsV1Write.writeWithV1(V1FallbackWriters.scala:79)
	at org.apache.spark.sql.execution.datasources.v2.SupportsV1Write.writeWithV1$(V1FallbackWriters.scala:78)
	at org.apache.spark.sql.execution.datasources.v2.OverwriteByExpressionExecV1.writeWithV1(V1FallbackWriters.scala:51)
	at org.apache.spark.sql.execution.datasources.v2.V1FallbackWriters.run(V1FallbackWriters.scala:66)
	at org.apache.spark.sql.execution.datasources.v2.V1FallbackWriters.run$(V1FallbackWriters.scala:65)
	at org.apache.spark.sql.execution.datasources.v2.OverwriteByExpressionExecV1.run(V1FallbackWriters.scala:51)
	at org.apache.spark.sql.execution.datasources.v2.V2CommandExec.result$lzycompute(V2CommandExec.scala:43)
	at org.apache.spark.sql.execution.datasources.v2.V2CommandExec.result(V2CommandExec.scala:43)
	at org.apache.spark.sql.execution.datasources.v2.V2CommandExec.executeCollect(V2CommandExec.scala:49)
	at org.apache.spark.sql.execution.QueryExecution$$anonfun$eagerlyExecuteCommands$1.$anonfun$applyOrElse$1(QueryExecution.scala:98)
	at org.apache.spark.sql.execution.SQLExecution$.$anonfun$withNewExecutionId$6(SQLExecution.scala:109)
	at org.apache.spark.sql.execution.SQLExecution$.withSQLConfPropagated(SQLExecution.scala:169)
	at org.apache.spark.sql.execution.SQLExecution$.$anonfun$withNewExecutionId$1(SQLExecution.scala:95)
	at org.apache.spark.sql.SparkSession.withActive(SparkSession.scala:779)
	at org.apache.spark.sql.execution.SQLExecution$.withNewExecutionId(SQLExecution.scala:64)
	at org.apache.spark.sql.execution.QueryExecution$$anonfun$eagerlyExecuteCommands$1.applyOrElse(QueryExecution.scala:98)
	at org.apache.spark.sql.execution.QueryExecution$$anonfun$eagerlyExecuteCommands$1.applyOrElse(QueryExecution.scala:94)
	at org.apache.spark.sql.catalyst.trees.TreeNode.$anonfun$transformDownWithPruning$1(TreeNode.scala:584)
	at org.apache.spark.sql.catalyst.trees.CurrentOrigin$.withOrigin(TreeNode.scala:176)
	at org.apache.spark.sql.catalyst.trees.TreeNode.transformDownWithPruning(TreeNode.scala:584)
	at org.apache.spark.sql.catalyst.plans.logical.LogicalPlan.org$apache$spark$sql$catalyst$plans$logical$AnalysisHelper$$super$transformDownWithPruning(LogicalPlan.scala:30)
	at org.apache.spark.sql.catalyst.plans.logical.AnalysisHelper.transformDownWithPruning(AnalysisHelper.scala:267)
	at org.apache.spark.sql.catalyst.plans.logical.AnalysisHelper.transformDownWithPruning$(AnalysisHelper.scala:263)
	at org.apache.spark.sql.catalyst.plans.logical.LogicalPlan.transformDownWithPruning(LogicalPlan.scala:30)
	at org.apache.spark.sql.catalyst.plans.logical.LogicalPlan.transformDownWithPruning(LogicalPlan.scala:30)
	at org.apache.spark.sql.catalyst.trees.TreeNode.transformDown(TreeNode.scala:560)
	at org.apache.spark.sql.execution.QueryExecution.eagerlyExecuteCommands(QueryExecution.scala:94)
	at org.apache.spark.sql.execution.QueryExecution.commandExecuted$lzycompute(QueryExecution.scala:81)
	at org.apache.spark.sql.execution.QueryExecution.commandExecuted(QueryExecution.scala:79)
	at org.apache.spark.sql.Dataset.<init>(Dataset.scala:219)
	at org.apache.spark.sql.Dataset$.$anonfun$ofRows$2(Dataset.scala:99)
	at org.apache.spark.sql.SparkSession.withActive(SparkSession.scala:779)
	at org.apache.spark.sql.Dataset$.ofRows(Dataset.scala:96)
	at org.apache.spark.sql.SparkSession.$anonfun$sql$1(SparkSession.scala:622)
	at org.apache.spark.sql.SparkSession.withActive(SparkSession.scala:779)
	at org.apache.spark.sql.SparkSession.sql(SparkSession.scala:617)
	at org.apache.spark.sql.SQLContext.sql(SQLContext.scala:651)
	at org.apache.spark.sql.hive.thriftserver.SparkSQLDriver.run(SparkSQLDriver.scala:67)
	at org.apache.spark.sql.hive.thriftserver.SparkSQLCLIDriver.processCmd(SparkSQLCLIDriver.scala:384)
	at org.apache.spark.sql.hive.thriftserver.SparkSQLCLIDriver.$anonfun$processLine$1(SparkSQLCLIDriver.scala:504)
	at org.apache.spark.sql.hive.thriftserver.SparkSQLCLIDriver.$anonfun$processLine$1$adapted(SparkSQLCLIDriver.scala:498)
	at scala.collection.Iterator.foreach(Iterator.scala:941)
	at scala.collection.Iterator.foreach$(Iterator.scala:941)
	at scala.collection.AbstractIterator.foreach(Iterator.scala:1429)
	at scala.collection.IterableLike.foreach(IterableLike.scala:74)
	at scala.collection.IterableLike.foreach$(IterableLike.scala:73)
	at scala.collection.AbstractIterable.foreach(Iterable.scala:56)
	at org.apache.spark.sql.hive.thriftserver.SparkSQLCLIDriver.processLine(SparkSQLCLIDriver.scala:498)
	at org.apache.spark.sql.hive.thriftserver.SparkSQLCLIDriver$.main(SparkSQLCLIDriver.scala:286)
	at org.apache.spark.sql.hive.thriftserver.SparkSQLCLIDriver.main(SparkSQLCLIDriver.scala)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.apache.spark.deploy.JavaMainApplication.start(SparkApplication.scala:52)
	at org.apache.spark.deploy.SparkSubmit.org$apache$spark$deploy$SparkSubmit$$runMain(SparkSubmit.scala:984)
	at org.apache.spark.deploy.SparkSubmit.doRunMain$1(SparkSubmit.scala:191)
	at org.apache.spark.deploy.SparkSubmit.submit(SparkSubmit.scala:214)
	at org.apache.spark.deploy.SparkSubmit.doSubmit(SparkSubmit.scala:90)
	at org.apache.spark.deploy.SparkSubmit$$anon$2.doSubmit(SparkSubmit.scala:1072)
	at org.apache.spark.deploy.SparkSubmit$.main(SparkSubmit.scala:1081)
	at org.apache.spark.deploy.SparkSubmit.main(SparkSubmit.scala)
Caused by: java.time.DateTimeException: For input string: '19831'.
	at org.apache.paimon.utils.BinaryStringUtils.toDate(BinaryStringUtils.java:281)
	at org.apache.paimon.utils.TypeUtils.castFromStringInternal(TypeUtils.java:141)
	at org.apache.paimon.utils.TypeUtils.castFromString(TypeUtils.java:76)
	at org.apache.paimon.predicate.PredicateBuilder.partition(PredicateBuilder.java:370)
	at org.apache.paimon.operation.FileStoreCommitImpl.overwrite(FileStoreCommitImpl.java:419)
	at org.apache.paimon.table.sink.TableCommitImpl.commitMultiple(TableCommitImpl.java:225)
	at org.apache.paimon.table.sink.TableCommitImpl.commit(TableCommitImpl.java:200)
	at org.apache.paimon.table.sink.TableCommitImpl.commit(TableCommitImpl.java:179)
	at org.apache.paimon.table.sink.TableCommitImpl.commit(TableCommitImpl.java:163)
	at org.apache.paimon.spark.commands.PaimonSparkWriter.commit(PaimonSparkWriter.scala:109)
```

### What doesn't meet your expectations?

write data fail

### Anything else?

I have reviewed latest source code of master , this exception is because of 
org.apache.paimon.spark.SparkFilterConverter#convertLiteral(int, java.lang.Object)
![image](https://github.com/apache/paimon/assets/51162281/00bec213-bb0f-47a8-9bc0-74fbc6a35945)
![image](https://github.com/apache/paimon/assets/51162281/7d6b394c-42b9-4582-94e5-93f34614e68c)
Paimon write converted type first, e.g: Date string convert to days between and midnight, January 1, 1970 UTC.

But org.apache.paimon.predicate.PredicateBuilder#partition  process all typ ecast to string using org.apache.paimon.utils.TypeUtils#castFromStringInternal
![image](https://github.com/apache/paimon/assets/51162281/10d9400c-1db3-4645-b4ef-9fcb73c62671)


I am trying to fix it in pr

![image](https://github.com/apache/paimon/assets/51162281/806787b1-2962-4b06-99cf-e957255c5af2)



### Are you willing to submit a PR?

- [X] I'm willing to submit a PR!

--- a/paimon-common/src/main/java/org/apache/paimon/utils/BinaryStringUtils.java
+++ b/paimon-common/src/main/java/org/apache/paimon/utils/BinaryStringUtils.java
@@ -276,7 +276,12 @@ public class BinaryStringUtils {
     }
 
     public static int toDate(BinaryString input) throws DateTimeException {
-        Integer date = DateTimeUtils.parseDate(input.toString());
+        String str = input.toString();
+        if (StringUtils.isNumeric(str)) {
+            // for Integer.toString conversion
+            return toInt(input);
+        }
+        Integer date = DateTimeUtils.parseDate(str);
         if (date == null) {
             throw new DateTimeException("For input string: '" + input + "'.");
         }
@@ -285,7 +290,12 @@ public class BinaryStringUtils {
     }
 
     public static int toTime(BinaryString input) throws DateTimeException {
-        Integer date = DateTimeUtils.parseTime(input.toString());
+        String str = input.toString();
+        if (StringUtils.isNumeric(str)) {
+            // for Integer.toString conversion
+            return toInt(input);
+        }
+        Integer date = DateTimeUtils.parseTime(str);
         if (date == null) {
             throw new DateTimeException("For input string: '" + input + "'.");
         }

--- a/paimon-common/src/main/java/org/apache/paimon/utils/StringUtils.java
+++ b/paimon-common/src/main/java/org/apache/paimon/utils/StringUtils.java
@@ -554,4 +554,17 @@ public class StringUtils {
     public static String caseSensitiveConversion(String str, boolean caseSensitive) {
         return caseSensitive ? str : str.toLowerCase();
     }
+
+    public static boolean isNumeric(final CharSequence cs) {
+        if (isEmpty(cs)) {
+            return false;
+        }
+        final int sz = cs.length();
+        for (int i = 0; i < sz; i++) {
+            if (!Character.isDigit(cs.charAt(i))) {
+                return false;
+            }
+        }
+        return true;
+    }
 }

--- a/paimon-spark/paimon-spark-common/src/test/scala/org/apache/paimon/spark/sql/InsertOverwriteTest.scala
+++ b/paimon-spark/paimon-spark-common/src/test/scala/org/apache/paimon/spark/sql/InsertOverwriteTest.scala
@@ -19,9 +19,10 @@
 package org.apache.paimon.spark.sql
 
 import org.apache.paimon.spark.PaimonSparkTestBase
-
 import org.apache.spark.sql.Row
 import org.apache.spark.sql.types._
+
+import java.sql.Date
 
 class InsertOverwriteTest extends PaimonSparkTestBase {
 
@@ -328,4 +329,19 @@ class InsertOverwriteTest extends PaimonSparkTestBase {
       }
   }
 
+  test(s"insert overwrite date type partition table") {
+    spark.sql(s"""
+                 |CREATE TABLE T (
+                 |  id STRING,
+                 |  dt date)
+                 |PARTITIONED BY (dt)
+                 |TBLPROPERTIES (
+                 |  'primary-key' = 'id,dt',
+                 |  'bucket' = '3'
+                 |);
+                 |""".stripMargin)
+
+    spark.sql("INSERT OVERWRITE T partition (dt='2024-04-18') values(1)")
+    checkAnswer(spark.sql("SELECT * FROM T"), Row("1", Date.valueOf("2024-04-18")))
+  }
 }

--- a/paimon-spark/paimon-spark-common/src/test/scala/org/apache/paimon/spark/sql/InsertOverwriteTest.scala
+++ b/paimon-spark/paimon-spark-common/src/test/scala/org/apache/paimon/spark/sql/InsertOverwriteTest.scala
@@ -19,6 +19,7 @@
 package org.apache.paimon.spark.sql
 
 import org.apache.paimon.spark.PaimonSparkTestBase
+
 import org.apache.spark.sql.Row
 import org.apache.spark.sql.types._
 

